Client/Helpers/TabEventHandler.swift

```diff
@@ -0,0 +1,147 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Storage
+
+/**
+ * A handler can be a plain old swift object. It does not need to extend any
+ * other object, but can.
+ *
+ * Handlers should register for tab events with the `registerFor` method, and
+ * cleanup with the `unregister` method.
+ *
+ * ```
+ * class HandoffHandler {
+ *     var tabObservers: TabObservers!
+ *
+ *     init() {
+ *         tabObservers = registerFor(.didLoadFavicon, .didLoadPageMetadata)
+ *     }
+ *
+ *     deinit {
+ *         unregister(tabObservers)
+ *     }
+ * }
+ * ```
+ *
+ * Handlers can implement any or all `TabEventHandler` methods. If you
+ * implement a method, you should probably `registerFor` the event above.
+ *
+ * ```
+ * extension HandoffHandler: TabEventHandler {
+ *     func tab(_ tab: Tab, didLoadPageMetadata metadata: PageMetadata) {
+ *         print("\(tab) has \(pageMetadata)")
+ *     }
+ *
+ *     func tab(_ tab: Tab, didLoadFavicon favicon: Favicon) {
+ *         print("\(tab) has \(favicon)")
+ *     }
+ * }
+ * ```
+ *
+ * Tab events should probably be only posted from one place, to avoid cycles.
+ *
+ * ```
+ * TabEvent.post(.didLoadPageMetadata(aPageMetadata), for: tab)
+ * ```
+ *
+ * In this manner, we are able to use the notification center and have type safety.
+ *
+ */
+// As we want more events we add more here.
+// Each event needs:
+// 1. a method in the TabEventHandler.
+// 2. a default implementation of the method – so not everyone needs to implement it
+// 3. a TabEventLabel, which is needed for registration
+// 4. a TabEvent, with whatever parameters are needed.
+//    i) a case to map the event to the event label (var label)
+//   ii) a case to map the event to the event handler (func handle:with:)
+protocol TabEventHandler {
+    func tab(_ tab: Tab, didLoadPageMetadata metadata: PageMetadata)
+    func tab(_ tab: Tab, didLoadFavicon favicon: Favicon)
+}
+
+// Provide default implmentations, because we don't want to litter the code with
+// empty methods, and `@objc optional` doesn't really work very well.
+extension TabEventHandler {
+    func tab(_ tab: Tab, didLoadPageMetadata metadata: PageMetadata) {}
+    func tab(_ tab: Tab, didLoadFavicon favicon: Favicon) {}
+}
+
+enum TabEventLabel: String {
+    case didLoadPageMetadata = "didLoadPageMetadata"
+    case didLoadFavicon = "didLoadFavicon"
```

Comment:
IIRC  `.rawValue` for String enum is the label name, so specifying string isn't needed.